File: apps/android/app/src/main/kotlin/io/palyra/mobile/companion/CompanionShell.kt
package io.palyra.mobile.companion

enum class MobileSurface {
    APPROVALS,
    INBOX,
    SESSIONS,
    VOICE_NOTE,
    SETTINGS,
}

data class CompanionShellState(
    val locale: String = "en",
    val surface: MobileSurface = MobileSurface.INBOX,
    val contract: MobileBootstrapContract = MobileBootstrapContract(),
    val cachedState: MobileCachedState = MobileCachedState(),
    val recentSessions: List<MobileSessionRecap> = emptyList(),
    val approvals: List<MobileApprovalSummary> = emptyList(),
    val offlineBannerVisible: Boolean = false,
    val revokeBannerVisible: Boolean = false,
)

sealed interface CompanionAction {
    data class SwitchSurface(val surface: MobileSurface) : CompanionAction
    data class SetLocale(val locale: String) : CompanionAction
    data class SetOfflineBanner(val visible: Boolean) : CompanionAction
    data class SetRevokeBanner(val visible: Boolean) : CompanionAction
    data class ReplaceApprovals(val approvals: List<MobileApprovalSummary>) : CompanionAction
    data class ReplaceRecentSessions(val recaps: List<MobileSessionRecap>) : CompanionAction
}

fun reduceCompanionShell(
    state: CompanionShellState,
    action: CompanionAction,
): CompanionShellState {
    return when (action) {
        is CompanionAction.SwitchSurface -> state.copy(surface = action.surface)
        is CompanionAction.SetLocale -> state.copy(locale = action.locale)
        is CompanionAction.SetOfflineBanner -> state.copy(offlineBannerVisible = action.visible)
        is CompanionAction.SetRevokeBanner -> state.copy(revokeBannerVisible = action.visible)
        is CompanionAction.ReplaceApprovals -> state.copy(approvals = action.approvals)
        is CompanionAction.ReplaceRecentSessions -> state.copy(recentSessions = action.recaps)
    }
}

File: apps/ios/Sources/MobileCompanion/CompanionShell.swift
import Foundation

enum MobileSurface: String, Codable {
    case approvals
    case inbox
    case sessions
    case voiceNote = "voice_note"
    case settings
}

struct CompanionShellState: Codable, Equatable {
    var locale: String = "en"
    var surface: MobileSurface = .inbox
    var contract: MobileBootstrapContract = .init()
    var cachedState: MobileCachedState = .init()
    var recentSessions: [MobileSessionRecap] = []
    var approvals: [MobileApprovalSummary] = []
    var offlineBannerVisible: Bool = false
    var revokeBannerVisible: Bool = false
}

enum CompanionAction: Equatable {
    case switchSurface(MobileSurface)
    case setLocale(String)
    case setOfflineBanner(Bool)
    case setRevokeBanner(Bool)
    case replaceApprovals([MobileApprovalSummary])
    case replaceRecentSessions([MobileSessionRecap])
}

func reduceCompanionShell(
    state: CompanionShellState,
    action: CompanionAction
) -> CompanionShellState {
    var next = state
    switch action {
    case let .switchSurface(surface):
        next.surface = surface
    case let .setLocale(locale):
        next.locale = locale
    case let .setOfflineBanner(visible):
        next.offlineBannerVisible = visible
    case let .setRevokeBanner(visible):
        next.revokeBannerVisible = visible
    case let .replaceApprovals(approvals):
        next.approvals = approvals
    case let .replaceRecentSessions(recaps):
        next.recentSessions = recaps
    }
    return next
}
